### ISSUE TITLE
Producer buffer grows indefinetly

### ISSUE THREAD

**Issue by mprahlkamps** [confluentinc/confluent-kafka-python#810]
Description
===========

When sending many messages from a single producer (1000 messages/second and more), the library has no problem sending those messages but the producer buffer grows indefinitely until a BufferError is raised. I have a consumer, receiving all messages in realtime.

```
2020-03-29 19:58:40,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 83923, Produced (since flush): 83923, Producer queue: 160030, Sent: 56 KB/s                             
2020-03-29 19:58:41,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 84922, Produced (since flush): 84922, Producer queue: 161939, Sent: 56 KB/s                             
2020-03-29 19:58:42,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 85922, Produced (since flush): 85922, Producer queue: 163828, Sent: 56 KB/s                             
2020-03-29 19:58:43,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 86921, Produced (since flush): 86921, Producer queue: 165725, Sent: 56 KB/s                             
2020-03-29 19:58:44,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 87920, Produced (since flush): 87920, Producer queue: 167619, Sent: 56 KB/s                             
2020-03-29 19:58:45,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 88919, Produced (since flush): 88919, Producer queue: 169521, Sent: 56 KB/s                             
2020-03-29 19:58:46,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 89918, Produced (since flush): 89918, Producer queue: 171436, Sent: 56 KB/s                             
2020-03-29 19:58:47,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 90916, Produced (since flush): 90916, Producer queue: 173328, Sent: 56 KB/s                             
2020-03-29 19:58:48,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 91915, Produced (since flush): 91915, Producer queue: 175238, Sent: 56 KB/s                             
2020-03-29 19:58:49,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 92913, Produced (since flush): 92913, Producer queue: 177148, Sent: 56 KB/s                             
2020-03-29 19:58:50,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 93912, Produced (since flush): 93912, Producer queue: 179051, Sent: 56 KB/s                             
2020-03-29 19:58:51,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 94911, Produced (since flush): 94911, Producer queue: 180948, Sent: 56 KB/s                             
2020-03-29 19:58:52,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 95910, Produced (since flush): 95910, Producer queue: 182845, Sent: 56 KB/s                             
2020-03-29 19:58:53,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 96909, Produced (since flush): 96909, Producer queue: 184766, Sent: 56 KB/s                             
2020-03-29 19:58:54,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 97908, Produced (since flush): 97908, Producer queue: 186664, Sent: 56 KB/s                             
2020-03-29 19:58:55,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 98907, Produced (since flush): 98907, Producer queue: 188583, Sent: 56 KB/s                             
2020-03-29 19:58:56,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 99906, Produced (since flush): 99906, Producer queue: 190483, Sent: 56 KB/s                             
2020-03-29 19:58:56,807 - 27688 kafka_performance_testing.producer.performance_producer - INFO - Flushing...                                                                                                       
2020-03-29 19:58:57,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 100875, Produced (since flush): 875, Producer queue: 1680, Sent: 54 KB/s
2020-03-29 19:58:58,713 - 27673 kafka_performance_testing.producer.performance_producer - INFO - Produced: 101874, Produced (since flush): 1874, Producer queue: 3578, Sent: 56 KB/s
```

The producer buffer seems to raise the error at a predictable amount of messages at around 190000-200000 messages.

A workaround to this is to catch the BufferError and flush the producer without any noticeable performance impact. 

confluent_kafka.version():
('1.3.0', 16973824)
confluent_kafka.libversion():
('1.3.0', 16974079)

Apache Kafka broker version:
2.4.1 (3 node cluster)

Producer configuration:
{"bootstrap.servers": ",".join(self._bootstrap_servers), "compression.type": "gzip"}

Operating System:
Debian GNU/Linux 10 (buster)

How to reproduce
================

Send 1000 (or more) messages per second. Monitor the producer buffer `len(producer)`.
Wait for BufferError and then flush.

**Comment by edenhill** [comment]
https://github.com/edenhill/librdkafka/blob/master/INTRODUCTION.md#high-throughput

TLDR: set `linger.ms=50` (or more) to increase batching and thus throughput.

Also make sure you call poll() frequently to serve delivery reports.

**Comment by mprahlkamps** [comment]
Since I do not have a throughput problem, linger time would not make any difference.

Calling poll() regularly did indeed fix the problem. Most of the examples I saw, mention poll as a trigger to serve callbacks.

What would be a good interval to poll? After each produced message? Every second?

Thanks

**Comment by edenhill** [comment]
Call poll() frequently enough so the internal queues don't fill up, the easiest way to do that is by calling poll() before or after each produce().
Additionally, if you have periods where there is nothing to produce you will also want to call poll() frequently to serve delivery reports from previous produces.
